{% extends "layout.html" %}

{% block header %}
  <h2><img src="{{ url_for('static', filename='images/navbar/background_reading.png') }}" alt="background reading page icon" class="logo"> {% block title %}Background Reading{% endblock title %}</h2>
{% endblock header %}

{% block content %}
  <p>
    <i>Just looking for what a certain word means?
    Check out the <a href="#glossary">Glossary</a> for a quick reference including many words and acronyms used throughout the website.</i>
  </p>

  <p>
    To start, we will need to know what whole genome duplication (WGD) is.
    WGD, as the name may suggest, is the duplication of the entire genome, either during meiosis or mitosis.
    In meiosis this is due to nondisjunction, when chromosomes fail to separate<sup><a href="#background-reading-ref-12">12</a></sup>, and in mitosis this is due to somatic doubling.<sup><a href="#background-reading-ref-11">11</a></sup>
    This duplication results in paralogous genes<sup><a href="#background-reading-ref-16">16</a></sup>, the complete set of these genes in a genome being called a paranome.<sup><a href="#background-reading-ref-15">15</a></sup>
    Though a lot is not (completely) known about WGD, it is clear that there is an increase in expression divergence, a phynotypic trait showing difference during evolution<sup><a href="#background-reading-ref-21">21</a></sup>;
    regulatory-motif divergence, a sequence which serves a regulatory function<sup><a href="#background-reading-ref-20">20</a></sup>;
    and coding sequence divergence.<sup><a href="#background-reading-ref-6">6</a>, <a href="#background-reading-ref-18">18</a></sup>
    WGD is common in plants, though its role has not yet been rigorously tested, there is a strong theoretical basis for it being a facilitator in diversification.<sup><a href="#background-reading-ref-2">2</a>, <a href="#background-reading-ref-3">3</a>, <a href="#background-reading-ref-5">5</a>, <a href="#background-reading-ref-10">10</a></sup>
  </p>

  <h2>WGD Tools</h2>
  <p>
    To aid in identifying and dating WGD events, there is a suite of tools, wgd.<sup><a href="#background-reading-ref-1">1</a>, <a href="#background-reading-ref-2">2</a></sup>
    To avoid confusion, WGD capitalised refers to whole genome duplication, while wgd lowercased refers to the wgd tool.
    The tool was first made by a small team of researchers, and more recently updated by another person doing their own research.
    This website uses the <a href="https://github.com/heche-psb/wgd">second version of the tool</a>.
  </p>
  <p>
  The tool exists of seven smaller tools, subtools, each subtool plays a role in visualising and the analysis of WGD events.
  The subtools are as follows: dmd, ksd, mix, syn, peak, focus, viz.
  Short explanations of each can be found below, extensive documentation can be found on the <a href="{{ url_for('tools.index') }}">tools page</a>.
  </p>

  <h3>dmd</h3>
  <p>
    This is the first of the subtools, used for whole paranome, global/local MRBHs, orthogroups, and phylogeny interference.
    To use this subtool, a (or multiple) FASTA file containing DNA coding sequences need to be provided.
    These FASTA files can be downloaded from databases, e.g. <a href="https://www.ncbi.nlm.nih.gov/genbank/">GenBank</a>.
    The output of this is a paralogous/orthologous gene family file, provided as <code>.tsv</code>, which will be used in the next subtool.
  </p>
  <figure>
    <img src="{{ url_for('static', filename='images/background_reading/wgd_dmd.svg') }}" alt="dmd workflow.">
    <figcaption>dmd workflow. <a href="https://wgdv2.readthedocs.io/en/latest/recipes.html">Source</a></figcaption>
  </figure>

  <h3>ksd</h3>
  <p>
    Ksd is used to construct a K<sub>s</sub> distribution, with K<sub>s</sub> being synonymous substitutions.
    A synonymous substitution is a change in the DNA sequence which codes for the same amino acid.
    This is what the tool uses to analyze WGD events.
  </p>
  <figure>
    <img src="{{ url_for('static', filename='images/background_reading/ksd_distribution.svg') }}" alt="ksd distribution visualised.">
    <figcaption>ksd distribution visualised. <a href="https://wgdv2.readthedocs.io/en/latest/recipes.html">Source</a></figcaption>
  </figure>

  <h3>mix</h3>
  <p>
    This is used to mixture module a Ks distribution.
    In the background it uses scikit-learn's mixture moduling.
    A mixture module is a probabilistic model in statistics which shows the presence of subpopulations.<sup><a href="#background-reading-ref-24">24</a></sup>
  </p>
  <figure>
    <img src="{{ url_for('static', filename='images/background_reading/mixture_model.svg') }}" alt="Mixture model.">
    <figcaption>Mixture model. <a href="https://wgdv2.readthedocs.io/en/latest/recipes.html">Source</a></figcaption>
  </figure>

  <h3>syn</h3>
  <p>
    Syn, short for synteny, is used for synteny inference.
    Synteny refers to (groups of) genes that are present more than once on the same chromosome.
  </p>
  <figure>
    <img src="{{ url_for('static', filename='images/background_reading/synteny.svg') }}" alt="Synteny inference showing the distribution of duplicated segments across chromosomes.">
    <figcaption>Synteny inference showing the distribution of duplicated segments across chromosomes. <a href="https://wgdv2.readthedocs.io/en/latest/recipes.html">Source</a></figcaption>
  </figure>

  <h3>peak</h3>
  <p>
    Peak is used to select anchor pairs in orthogroup construction.
  </p>
  <figure>
    <img src="{{ url_for('static', filename='images/background_reading/peak.svg') }}" alt="peak.">
    <figcaption>peak. <a href="https://wgdv2.readthedocs.io/en/latest/recipes.html">Source</a></figcaption>
  </figure>

  <h3>focus</h3>
  <p>
    Used for concatenation-/coalescence-based phylogenetic inference, creating functional annotation of gene families, and date WGD events.
  </p>
  <figure>
    <img src="{{ url_for('static', filename='images/background_reading/concatenation_vs_coalescence.png') }}" alt="Methods of concatenation and coalescence.">
    <figcaption>
      Methods of concatenation and coalescence.
      (A) Concatenation method joins all gene sequences of each species into each super-gene sequence.
      These super-gene sequences are then used in aligning and building phylogeny trees;
      (B) Coalescence method extracts the same gene sequences from all species and merges them into each single parallel-gene sequence file.
      Then all parallel-gene sequence files are aligned separately to build trees.
      Finally, all those trees are computed to get the final tree.
      <a href="https://www.researchgate.net/figure/Methods-of-concatenation-and-coalescence-A-Concatenation-method-joins-all-gene_fig2_357352303">Source</a>
    </figcaption>
  </figure>

  <h3>viz</h3>
  <p>
    This is the visualisation module of the tool used for K<sub>s</sub> distributions and synteny.
  </p>
  <figure>
    <img src="{{ url_for('static', filename='images/background_reading/viz.svg') }}" alt="K<sub>s</sub> distribution of multiple multiple species.">
    <figcaption>K<sub>s</sub> distribution of multiple multiple species. <a href="https://wgdv2.readthedocs.io/en/latest/recipes.html">Source</a></figcaption>
  </figure>

  <section aria-labelledby="glossary">
    <h2 id="glossary">Glossary</h2>

    <ul>
      <li>WGD = Whole genome duplication</li>
      <li>Polyploidization = Synonym for WGD.</li>
      <li>Wgd = Refers to the whole genome duplication tool.</li>
      <li>Nondisjunction = When chromosomes fail to separate during the cell cycle, causing too many chromosomes to be present in the daughter cells.</li>
      <li>Somatic doubling = Doubling of the amount of chromosomes in a new cell.<sup><a href="#background-reading-ref-19">19</a></sup></li>
      <li>FASTA = File format used to store one or more sequences (DNA or RNA).</li>
      <li>RBH = Reciprocal best hits, used to compare genomes in orthology. When two genes in different genomes find each other as the best match.<sup><a href="#background-reading-ref-4">4</a></sup></li>
      <li>Paranome = Complete set of paralogous genes in a genome.<sup><a href="#background-reading-ref-15">15</a></sup></li>
      <li>Paralogues/Paralogous genes = Genes separated by gene duplication events.</li>
      <li>Orthology = Relationship between multiple genes that share a common ancestor.<sup><a href="#background-reading-ref-17">17</a></sup></li>
      <li>dmd = A subtool within the wgd tool, used for whole paranome, global/local MRBHs, orthogroups, and phylogeny interference.<sup><a href="#background-reading-ref-1">1</a></sup></li>
      <li>Ks distribution = Distribution which denotes expected number of synonymous substitutions per site between two DNA sequences.</li>
      <li>Synonymous substitution = Within context of DNA and RNA, it's a substition in DNA which codes for the same amino acid.</li>
      <li>Ksd = A subtool within the wgd tool, used to construct the Ks distribution.<sup><a href="#background-reading-ref-1">1</a></sup></li>
      <li>mix = A subtool within the wgd tool, used to mixture module a Ks distribution.<sup><a href="#background-reading-ref-1">1</a></sup></li>
      <li>Synteny = When (group of) genes are present more than once on the same chromosome.</li>
      <li>Syn = A subtool within the wgd tool, used for synteny inference.<sup><a href="#background-reading-ref-1">1</a></sup></li>
      <li>WGD dating = The dating of WGD events.</li>
      <li>Peak = A subtool within the wgd tool, used to find anchor pairs used in WGD dating.<sup><a href="#background-reading-ref-1">1</a></sup></li>
      <li>Concatenation = The joining of gene sequences per species.<sup><a href="#background-reading-ref-14">14</a></sup></li>
      <li>Coalescence = Extracts same gene sequences from all the species and merges them.<sup><a href="#background-reading-ref-14">14</a></sup></li>
      <li>focus = A subtool within the wgd tool, used for concatenation-/coalescence-based phylogenetic inference, creating functional annotation of gene families, and date WGD events.<sup><a href="#background-reading-ref-1">1</a></sup></li>
      <li>viz = A subtool within the wgd tool, used to visualise Ks distributions and synteny.<sup><a href="#background-reading-ref-1">1</a></sup></li>
    </ul>
  </section>

  <section aria-labelledby="references">
    <h2 id="references">References</h2>

    <ol>
      <li id="background-reading-ref-1">
        “wgd v2 documentation — wgd v2 v2 documentation.” <a href="https://wgdv2.readthedocs.io/en/latest/">https://wgdv2.readthedocs.io/en/latest/</a>
      </li>
      <li id="background-reading-ref-2">
        M. Sémon and K. H. Wolfe, “Consequences of genome duplication,” Current Opinion in Genetics & Development, vol. 17, no. 6, pp. 505–512, Nov. 2007, doi: <a href="https://www.doi.org/10.1016/j.gde.2007.09.007">10.1016/j.gde.2007.09.007</a>.
      </li>
      <li id="background-reading-ref-3">
        D. Yu et al., “Hagfish genome elucidates vertebrate whole-genome duplication events and their evolutionary consequences,” Nature Ecology & Evolution, vol. 8, no. 3, pp. 519–535, Jan. 2024, doi: <a href="https://www.doi.org/10.1038/s41559-023-02299-z">10.1038/s41559-023-02299-z</a>.
      </li>
      <li id="background-reading-ref-4">
        N. Ward and G. Moreno-Hagelsieb, “Quickly Finding Orthologs as Reciprocal Best Hits with BLAT, LAST, and UBLAST: How Much Do We Miss?,” PLoS ONE, vol. 9, no. 7, p. e101850, Jul. 2014, doi: <a href="https://www.doi.org/10.1371/journal.pone.0101850">10.1371/journal.pone.0101850</a>.
      </li>
      <li id="background-reading-ref-5">
        S. P. Otto, “The evolutionary consequences of polyploidy,” Cell, vol. 131, no. 3, pp. 452–462, Nov. 2007, doi: <a href="https://www.doi.org/10.1016/j.cell.2007.10.022">10.1016/j.cell.2007.10.022</a>.
      </li>
      <li id="background-reading-ref-6">
        L. Carretero‐Paulet and Y. Van De Peer, “The evolutionary conundrum of whole‐genome duplication,” American Journal of Botany, vol. 107, no. 8, pp. 1101–1105, Aug. 2020, doi: <a href="https://www.doi.org/10.1002/ajb2.1520">10.1002/ajb2.1520</a>.
      </li>
      <li id="background-reading-ref-7">
        F. Marlétaz et al., “The hagfish genome and the evolution of vertebrates,” bioRxiv (Cold Spring Harbor Laboratory), Apr. 2023, doi: <a href="https://www.doi.org/10.1101/2023.04.17.537254">10.1101/2023.04.17.537254</a>.
      </li>
      <li id="background-reading-ref-8">
        A. Zwaenepoel and Y. Van De Peer, “wgd—simple command line tools for the analysis of ancient whole-genome duplications,” Bioinformatics, vol. 35, no. 12, pp. 2153–2155, Nov. 2018, doi: <a href="https://www.doi.org/10.1093/bioinformatics/bty915">10.1093/bioinformatics/bty915</a>.
      </li>
      <li id="background-reading-ref-9">
        H. Chen, A. Zwaenepoel, and Y. Van De Peer, “wgd v2: a suite of tools to uncover and date ancient polyploidy and whole-genome duplication,” Bioinformatics, vol. 40, no. 5, Apr. 2024, doi: <a href="https://www.doi.org/10.1093/bioinformatics/btae272">10.1093/bioinformatics/btae272</a>.
      </li>
      <li id="background-reading-ref-10">
        J. W. Clark and P. C. J. Donoghue, “Whole-Genome duplication and plant macroevolution,” Trends in Plant Science, vol. 23, no. 10, pp. 933–945, Aug. 2018, doi: <a href="https://www.doi.org/10.1016/j.tplants.2018.07.006">10.1016/j.tplants.2018.07.006</a>.
      </li>
      <li id="background-reading-ref-11">
        S. M. K. Glasauer and S. C. F. Neuhauss, “Whole-genome duplication in teleost fishes and its evolutionary consequences,” Molecular Genetics and Genomics, vol. 289, no. 6, pp. 1045–1060, Aug. 2014, doi: <a href="https://www.doi.org/10.1007/s00438-014-0889-2">10.1007/s00438-014-0889-2</a>.
      </li>
      <li id="background-reading-ref-12">
        S. F. Gottlieb, C. Tupper, C. C. Kerndt, and D. H. Tegay, “Genetics, nondisjunction,” StatPearls - NCBI Bookshelf, Aug. 14, 2023. <a href="https://www.ncbi.nlm.nih.gov/books/NBK482240/">https://www.ncbi.nlm.nih.gov/books/NBK482240/</a>
      </li>
      <li id="background-reading-ref-13">
        J. Luo, J. Chen, W. Guo, Z. Yang, K.-J. Lim, and Z. Wang, “Reassessment of Annamocarya sinesis (Carya sinensis) Taxonomy through Concatenation and Coalescence Phylogenetic Analysis,” Plants, vol. 11, no. 1, p. 52, Dec. 2021, doi: <a href="https://www.doi.org/10.3390/plants11010052">10.3390/plants11010052</a>.
      </li>
      <li id="background-reading-ref-14">
        J. Luo, J. Chen, W. Guo, Z. Yang, K.-J. Lim, and Z. Wang, “Reassessment of Annamocarya sinesis (Carya sinensis) Taxonomy through Concatenation and Coalescence Phylogenetic Analysis,” Plants, vol. 11, no. 1, p. 52, Dec. 2021, doi: <a href="https://www.doi.org/10.3390/plants11010052">10.3390/plants11010052</a>.
      </li>
      <li id="background-reading-ref-15">
        “paranome - Wiktionary, the free dictionary,” Wiktionary. <a href="https://en.wiktionary.org/wiki/paranome">https://en.wiktionary.org/wiki/paranome</a>
      </li>
      <li id="background-reading-ref-16">
        U.S. National Library of Medicine, “Homology: Orthologs and Paralogs.” <a href="https://www.nlm.nih.gov/ncbi/workshops/2023-08_BLAST_evol/ortho_para.html">https://www.nlm.nih.gov/ncbi/workshops/2023-08_BLAST_evol/ortho_para.html</a>
      </li>
      <li id="background-reading-ref-17">
        P. Van Baarlen, H. P. Van Esse, R. J. Siezen, and B. P. H. J. Thomma, “Challenges in plant cellular pathway reconstruction based on gene expression profiling,” Trends in Plant Science, vol. 13, no. 1, pp. 44–50, Dec. 2007, doi: <a href="https://www.doi.org/10.1016/j.tplants.2007.11.003">10.1016/j.tplants.2007.11.003</a>.
      </li>
      <li id="background-reading-ref-18">
        W.-H. Li, J. Yang, and X. Gu, “Expression divergence between duplicate genes,” Trends in Genetics, vol. 21, no. 11, pp. 602–607, Sep. 2005, doi: <a href="https://www.doi.org/10.1016/j.tig.2005.08.006">10.1016/j.tig.2005.08.006</a>.
      </li>
      <li id="background-reading-ref-19">
        “somatic doubling definition.” <a href="https://groups.molbiosci.northwestern.edu/holmgren/Glossary/Definitions/Def-S/somatic_doubling.html">https://groups.molbiosci.northwestern.edu/holmgren/Glossary/Definitions/Def-S/somatic_doubling.html</a>
      </li>
      <li id="background-reading-ref-20">
        Libretexts, “4.7: Regulatory motifs,” Biology LibreTexts, Mar. 17, 2021. <a href="https://bio.libretexts.org/Bookshelves/Computational_Biology/Book%3A_Computational_Biology_-_Genomes_Networks_and_Evolution_(Kellis_et_al.)/04%3A_Comparative_Genomics_I-_Genome_Annotation/4.07%3A_Regulatory_Motifs">https://bio.libretexts.org/Bookshelves/Computational_Biology/Book%3A_Computational_Biology_-_Genomes_Networks_and_Evolution_(Kellis_et_al.)/04%3A_Comparative_Genomics_I-_Genome_Annotation/4.07%3A_Regulatory_Motifs</a>
      </li>
      <li id="background-reading-ref-21">
        G. Glazko and A. Mushegian, “Measuring gene expression divergence: the distance to keep,” Biology Direct, vol. 5, no. 1, p. 51, Jan. 2010, doi: <a href="https://www.doi.org/10.1186/1745-6150-5-51">10.1186/1745-6150-5-51</a>.
      </li>
      <li id="background-reading-ref-22">
        Heche-Psb, “GitHub - heche-psb/wgd: wgd v2: a suite of tools to uncover and date ancient polyploidy and whole-genome duplication,” GitHub. <a href="https://github.com/heche-psb/wgd">https://github.com/heche-psb/wgd</a>
      </li>
      <li id="background-reading-ref-23">
        “GenBank Overview.” <a href="https://www.ncbi.nlm.nih.gov/genbank/">https://www.ncbi.nlm.nih.gov/genbank/</a>
      </li>
      <li id="background-reading-ref-24">
        Wikipedia contributors, “Mixture model,” Wikipedia, Feb. 17, 2025. <a href="https://en.wikipedia.org/wiki/Mixture_model">https://en.wikipedia.org/wiki/Mixture_model</a>
      </li>
    </ol>
  </section>

{% endblock content %}
